driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
>I want to use same pattern as possible for Ping, QueryContext() and other APIs.

I see what you concern. How about still letting normal connection reuse agagin but replacing line [72](https://github.com/go-sql-driver/mysql/pull/572/files#diff-7cd962f1d3aee9adf93d5c17b49f4530R72)~80 with these:
``` 
if err == nil {
  return nil
}
if mc.netConn != nil {
  errClose := mc.netConn.Close()
  if errClose != nil {
    errLog.Print(errClose)
  }
  mc.netConn = nil
}
return err
```